cedar-dafny/validation/typechecker.dfy

```diff
@@ -381,75 +389,98 @@ module validation.typechecker {
               if b then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
-          case (PrimitiveLit(EntityUID(u)),Set(_)) =>
-            match tryGetEUIDs(e2) {
-              case Some(euids) =>
-                // If the entity literal is an action, then use acts.descendantOfSet.
-                // Otherwise, use ets.possibleDescendantOfSet.
-                if isAction(u.ty)
-                then
-                  if acts.descendantOfSet(u,euids) then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
-                else
+            case (Var(v),Set(_)) =>
+              var et := getPrincipalOrResource(v);
+              match tryGetEUIDs(e2) {
+                case Some(euids) =>
                   var es := set euid <- euids :: euid.ty;
-                  var b := ets.possibleDescendantOfSet(u.ty,es);
+                  var b := et.None? || ets.possibleDescendantOfSet(et.value,es);
                   if b then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
-              case None => Ok(Type.Bool(AnyBool))
-            }
-          // otherwise, the result is unknown so return Bool
-          case _ => Ok(Type.Bool(AnyBool))
-        }
+                case None => Ok(Type.Bool(AnyBool))
+              }
+            // LHS is entity literal (or action, per above)
+            case (PrimitiveLit(EntityUID(u1)),PrimitiveLit(EntityUID(u2))) =>
+              // If the entity literal is an action, then use acts.descendantOf.
+              // Otherwise, use ets.possibleDescendantOf.
+              if isAction(u1.ty)
+              then
+                if acts.descendantOf(u1,u2) then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
+              else
+                var b := ets.possibleDescendantOf(u1.ty,u2.ty);
+                if b then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
+            case (PrimitiveLit(EntityUID(u)),Set(_)) =>
+              match tryGetEUIDs(e2) {
+                case Some(euids) =>
+                  // If the entity literal is an action, then use acts.descendantOfSet.
+                  // Otherwise, use ets.possibleDescendantOfSet.
+                  if isAction(u.ty)
+                  then
+                    if acts.descendantOfSet(u,euids) then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
+                  else
+                    var es := set euid <- euids :: euid.ty;
+                    var b := ets.possibleDescendantOfSet(u.ty,es);
+                    if b then Ok(Type.Bool(AnyBool)) else Ok(Type.Bool(False))
+                case None => Ok(Type.Bool(AnyBool))
+              }
+            // otherwise, the result is unknown so return Bool
+            case _ => Ok(Type.Bool(AnyBool))
+          };
+
+      if this.mode.isStrict() && outTy != Type.Bool(True) && outTy != Type.Bool(False)
+      then
+        var ety1 :- extractEntityType(elub1);
+        var ety2 :- extractEntityType(elub2.value);
+        if ets.possibleDescendantOf(ety1, ety2)
```

Comment:
I'm not sure what this extra check is buying. Aren't the cases already handled by the code above? If not, it would be nice to add a comment explaining why.